### PR TITLE
grt: first step to td-gr

### PR DESCRIPTION
**Pull request by eder-matheus**
This PR updates grt to have the capability to set layer ranges for timing-critical nets. At this stage, grt considers a net as "timing-critical" based on its negative slack values. A set of configuration options are added to the Tcl interface, and they are described in grt README.

### CODE REVIEW

File: src/grt/src/GlobalRouter.cpp [hunk 1]
```diff
@@ -304,6 +312,105 @@ void GlobalRouter::estimateRC()
...
 
+bool GlobalRouter::findWorstSlackNets(float worst_nets_percentage)
+{
+  std::vector<float> slack_values;
+  std::map<odb::dbNet*, float> slack_per_net;
+
+  rsz_->resizeSlackPreamble();
+  rsz_->findResizeSlacks();
+
+  std::vector<odb::dbNet*> worst_slack_nets = rsz_->resizeWorstSlackDbNets();
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:323]: Why are you using the virtual resizer API grt - resizing should already have happened.   Is this for inside of gpl?

File: src/grt/src/GlobalRouter.cpp [hunk 2]
```diff
@@ -304,6 +312,105 @@ void GlobalRouter::estimateRC()
...
+    return false;
+  }
+
+  for (Net& net : *nets_) {
+    float net_slack = rsz_->resizeNetSlack(net.getDbNet());
+    slack_values.push_back(net_slack);
+    slack_per_net[net.getDbNet()] = net_slack;
+  }
+
+  if (!slack_values.empty()) {
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:344]: Wouldn't this imply no nets at all?

File: src/grt/src/GlobalRouter.cpp [hunk 3]
```diff
@@ -304,6 +312,105 @@ void GlobalRouter::estimateRC()
...
+
+    int slack_idx = (slack_values.size() * worst_nets_percentage) - 1;
+    float max_slack = (max_negative_slack_ != 0) ? 
+                      max_negative_slack_ : slack_values[slack_idx];
+
+    int critical_nets_count = 0;
+    for (Net& net : *nets_) {
+      odb::Rect bbox = computeNetBBox(net);
+      if (slack_per_net[net.getDbNet()] < 0 &&
+          slack_per_net[net.getDbNet()] <= max_slack &&
```
Comment (maliberty) [inline src/grt/src/GlobalRouter.cpp:355]: these checks seems redundant, certainly the lookup is.

**Review by maliberty — COMMENTED** [review]
This only seems to set the net critical flag without actually using it anywhere.  Is that what "first step" means?

**Comment by maliberty** [comment]
@eder-matheus this seems to have gotten lost.  Is this still relevant?

**Comment by eder-matheus** [comment]
> @eder-matheus this seems to have gotten lost. Is this still relevant?

I will redo what I've tried here in another branch, following your comments. But I think this PR can be closed.

**Pull request closed without merging**